### DISCUSSION TITLE
Retrying a scenario only under certain conditions

### DISCUSSION THREAD

**Discussion by tgsmith61591** [cucumber-rs/cucumber#281] [Q&A] [answered]
I am reading [this section](https://cucumber-rs.github.io/cucumber/main/writing/retries.html) on `retry` in the documentation. I have a `given` step that fetches a cached token to use for a downstream API call. This token has an associated TTL, and must be refreshed occasionally. In the event that the step gets a stale token, my test will return a 403, in which event I'd like to retry the scenario.

My question is how I can amend the `retry` behavior to **only** retry for a certain event? Something like:

```gherkin
  @retry.for_http(403)
  Scenario: API call works
    Given I have a token  # get cached token, or refresh it if it has expired
    When I file a cucumber-rs issue
     Then I will figure out how to solve my problem
```

Any ideas how I can approach this problem?

**Comment by ilslv** [comment, marked as answer]
@tgsmith61591 we don't support more complex logic for the retries and I don't think that we will ever do that. Quoting the `cucumber` book

> __WARNING__: Although this feature is supported, we highly recommend to use it as the last resort only. First, consider implementing in-[step](https://cucumber.io/docs/gherkin/reference#steps) retries with your own needs (like [exponential backoff](https://en.wikipedia.org/wiki/Exponential_backoff)). Other ways of dealing with flaky tests include, but not limited to: reducing number of concurrently executed scenarios (maybe even using `@serial` [tag](https://cucumber.io/docs/cucumber/api#tags)), mocking external environment, [controlling time in tests](https://docs.rs/tokio/1.0/tokio/time/fn.pause.html) or even [simulation testing](https://github.com/madsys-dev/madsim). It's always better to move towards tests determinism, rather than trying to tame their flakiness.

So I think the best way to deal with your situation is to implement in-step retry or some way to preemptively rotate API keys.
